### PR TITLE
Backend rspec

### PR DESCRIPTION
**Pull request by okeren-ww**
Added rspec tests for restaurant model and controller as an example. Also added a short helper module with reusable code.
Model & controller tests are straightforward and use Faker factories. They test the model structure & two controller methods (show & index)

### CODE REVIEW

File: spec/factories/restaurants.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+FactoryBot.define do
+  factory :restaurant do
+    name 'McDonalds'
```
Comment (barakcoh) [inline spec/factories/restaurants.rb]: use `Faker` as much as possible to expose input edge cases

File: spec/factories/restaurants.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    address 'Ibn Gabirol 30'
+    max_delivery_time 15
+    accepts_ten_bis true
+    cuisine_id 1
```
Comment (barakcoh) [inline spec/factories/restaurants.rb]: shouldn't the `Restaurant` `belong_to cuisine`? then the factory should have an association to the `cuisine`

File: spec/factories/reviews.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :review do
+    reviewer_name { Faker::SiliconValley.character }
+    rating { Faker::Number.between(0, 5) }
+    comment { Faker::ChuckNorris.fact }
```
Comment (barakcoh) [inline spec/factories/reviews.rb:18]: 😂

File: spec/models/Restaurant_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  # Model member verification
+  it { should validate_presence_of :name }
+  it { should validate_presence_of :cuisine_id }
+  it { should validate_presence_of :accepts_ten_bis }
+  it { should validate_presence_of :address }
+  it { should validate_presence_of :max_delivery_time }
+  it { should have_many(:reviews) }
+
+  context 'Restaurant with no parameters' do
+    it 'Should Raise 7 errors' do
```
Comment (barakcoh) [inline spec/models/Restaurant_spec.rb]: this is pretty non-descriptive. if tomorrow someone adds another validation and mistakingly removes an existing one this test will still pass

**Review by barakcoh — APPROVED** [review]

File: spec/controllers/Restaurant_spec.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+require 'rails_helper'
+
+describe RestaurantsController, type: :controller do
+  ratings = []
```
Comment (RonaTross) [inline spec/controllers/Restaurant_spec.rb]: setup shouldn't be done directly on the describe block, but rather in either a `let` or a `before` block. Specifically here, I'd create the random ratings inside restaurant_with_reviews factory (and still allow it to be overridden with a rating array passed in from the outside like you have now)

File: spec/controllers/Restaurant_spec.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  (0..4).each do |_i|
+    ratings.push Faker::Number.between(0, 5)
+  end
+
+  let!(:restaurant_with_reviews) { FactoryBot.create(:restaurant_with_reviews, dummy_ratings: ratings) }
+  let!(:restaurant_no_reviews) { FactoryBot.create(:restaurant) }
+
+  describe 'Get index' do
```
Comment (RonaTross) [inline spec/controllers/Restaurant_spec.rb:9]: minor: I think the convention is GET (all in upper case)

File: spec/factories/reviews.rb (hunk 3 above)
Comment (RonaTross) [reply spec/factories/reviews.rb:18]: OMG! this should go straight to TIL

File: spec/factories/reviews.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :review do
+    reviewer_name { Faker::SiliconValley.character }
```
Comment (RonaTross) [inline spec/factories/reviews.rb:16]: wut

File: spec/models/Restaurant_spec.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+require 'faker'
```
Comment (RonaTross) [inline spec/models/Restaurant_spec.rb]: Is this necessary?

File: spec/models/Restaurant_spec.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+describe Restaurant do
+  # Model member verification
+  it { should validate_presence_of :name }
```
Comment (RonaTross) [inline spec/models/Restaurant_spec.rb]: new syntax is `is_expected.to`

File: spec/controllers/Restaurant_spec.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    before do
+      get :index
+    end
+
+    it 'Returns 200 OK' do
+      expect(response).to have_http_status(200)
```
Comment (shem8) [inline spec/controllers/Restaurant_spec.rb:15]: `have_http_status(:ok)`

File: spec/controllers/Restaurant_spec.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      parsed_body = JSON.parse(response.body)
+      expect(parsed_body.size).to be(2)
+      compare_restaurants parsed_body.first.symbolize_keys, restaurant_with_reviews
+      compare_restaurants parsed_body.second.symbolize_keys, restaurant_no_reviews
+    end
+  end
+
+  describe 'Get show' do
+    before do
+      get :show, params: { id: restaurant_with_reviews.id }
```
Comment (shem8) [inline spec/controllers/Restaurant_spec.rb]: What about no reviews?

File: spec/models/Restaurant_spec.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  context 'Restaurant with no reviews' do
+    let!(:restaurant_no_reviews) { FactoryBot.create(:restaurant) }
+    it 'Should be created successfully and rating is 0' do
+      expect(restaurant_no_reviews.rating).to eql(0)
+    end
+  end
+
+  context 'Restaurant with reviews' do
+    ratings = []
+    (0..4).each do |_i|
```
Comment (shem8) [inline spec/models/Restaurant_spec.rb]: Put this logic in the factory

File: spec/models/Restaurant_spec.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  context 'Restaurant with no parameters' do
+    it 'Should Raise 7 errors' do
+      restaurant = Restaurant.create
+      expect(restaurant.errors.size).to eq(7)
+    end
+  end
+
+  context 'Restaurant with no reviews' do
+    let!(:restaurant_no_reviews) { FactoryBot.create(:restaurant) }
+    it 'Should be created successfully and rating is 0' do
```
Comment (RonaTross) [inline spec/models/Restaurant_spec.rb]: minor: missing an empty line after the `let`

File: spec/support/json_helper.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+module JsonHelper
+  def compare_restaurants(json_body, expected_restaurant)
```
Comment (RonaTross) [inline spec/support/json_helper.rb]: JsonHelper is a too general - I'd actually keep this insde the restaurant_spec file, and use a custom matcher here (https://relishapp.com/rspec/rspec-expectations/v/2-4/docs/custom-matchers/define-matcher)

File: app/models/restaurant.rb [hunk 15]
```diff
@@ -1,8 +1,10 @@
 class Restaurant < ApplicationRecord
   has_many :reviews
   validates :name, :cuisine_id, :rating, :accepts_ten_bis, :address, :max_delivery_time, presence: true
-  validates :rating, numericality: [greater_than_or_equal_to: 0, less_than_or_equal_to: 3]
+  validates :rating, numericality: [greater_than_or_equal_to: 0, less_than_or_equal_to: 5]
```
Comment (yossis) [inline app/models/restaurant.rb]: 👍

File: Gemfile [hunk 16]
```diff
@@ -42,7 +42,11 @@ group :development, :test do
   gem 'byebug', platforms: %i(mri mingw x64_mingw)
   # Adds support for Capybara system testing and selenium driver
   gem 'capybara', '~> 2.13'
+  gem 'factory_bot_rails'
+  gem 'faker', '~> 1.8', '>= 1.8.7'
+  gem 'rspec-rails', '~> 3.7'
   gem 'selenium-webdriver'
```
Comment (yossis) [inline Gemfile:49]: add annotate gem so you can see the schema in each model
https://github.com/ctran/annotate_models

File: app/controllers/restaurants_controller.rb [hunk 17]
```diff
@@ -27,7 +27,10 @@ def update
     render json: @restaurant
   end
 
-  def show; end
+  def show
+    set_restaurant
```
Comment (yossis) [inline app/controllers/restaurants_controller.rb]: Why do you use this and not `before_action`?

**Review by yossis — COMMENTED** [review]
great job

File: spec/models/Restaurant_spec.rb (hunk 4 above)
Comment (okeren-ww) [reply spec/models/Restaurant_spec.rb]: Cool, changed it to check for raising of RecordInvalid error

File: spec/factories/restaurants.rb (hunk 2 above)
Comment (okeren-ww) [reply spec/factories/restaurants.rb]: Correct, I was certain I did it, but looks like I didn't

File: Gemfile (hunk 16 above)
Comment (okeren-ww) [reply Gemfile:49]: really cool thx!

File: app/controllers/restaurants_controller.rb (hunk 17 above)
Comment (okeren-ww) [reply app/controllers/restaurants_controller.rb]: No particular reason. Changed it to before_action

**Comment by okeren-ww** [comment]
Fixed all issues

**Pull request merged**